File: apps/api/src/services/job.service.ts
import type { HydratedDocument } from "mongoose";
import type { CreateJobInput, DatePostedFilter, Job, JobFilterOptions, JobSearchFilters, UpdateJobInput } from "@believe-ai/shared";
import { MAX_PAGE_SIZE } from "@believe-ai/shared";
import type { JobDocument, SavedJobDocument } from "@believe-ai/server";
import { jobRepository } from "../repositories/job.repository.js";
import { savedJobRepository } from "../repositories/savedJob.repository.js";
import { searchExternalJobs } from "./jsearch.client.js";
import { NotFoundError, ValidationError } from "../errors/AppError.js";

const DATE_POSTED_MS: Record<Exclude<DatePostedFilter, "any">, number> = {
  "24h": 24 * 60 * 60 * 1000,
  "7d": 7 * 24 * 60 * 60 * 1000,
  "30d": 30 * 24 * 60 * 60 * 1000,
};

function datePostedToCutoff(datePosted: DatePostedFilter | undefined): Date | undefined {
  if (!datePosted || datePosted === "any") return undefined;
  return new Date(Date.now() - DATE_POSTED_MS[datePosted]);
}

function meetsSalaryFloor(job: Job, floor: number | undefined): boolean {
  if (floor == null) return true;
  const ceiling = job.salaryMax ?? job.salaryMin;
  return ceiling != null && ceiling >= floor;
}

function toDto(doc: HydratedDocument<JobDocument>): Job {
  return {
    id: doc._id.toString(),
    source: "internal",
    postedBy: doc.postedBy.toString(),
    title: doc.title,
    company: doc.company,
    location: doc.location ?? null,
    description: doc.description,
    skills: doc.skills ?? [],
    employmentType: (doc.employmentType as Job["employmentType"]) ?? null,
    workMode: (doc.workMode as Job["workMode"]) ?? null,
    experienceLevel: (doc.experienceLevel as Job["experienceLevel"]) ?? null,
    salaryMin: doc.salaryMin ?? null,
    salaryMax: doc.salaryMax ?? null,
    recruiterLinkedIn: doc.recruiterLinkedIn ?? null,
    applyUrl: doc.applyUrl ?? null,
    createdAt: doc.createdAt.toISOString(),
    updatedAt: doc.updatedAt.toISOString(),
  };
}

function toSavedDto(doc: HydratedDocument<SavedJobDocument>): Job {
  const s = doc.snapshot!;
  return {
    id: doc.jobId,
    source: s.source as Job["source"],
    postedBy: null,
    title: s.title,
    company: s.company,
    location: s.location ?? null,
    description: s.description ?? "",
    skills: s.skills ?? [],
    employmentType: (s.employmentType as Job["employmentType"]) ?? null,
    workMode: (s.workMode as Job["workMode"]) ?? null,
    experienceLevel: (s.experienceLevel as Job["experienceLevel"]) ?? null,
    salaryMin: s.salaryMin ?? null,
    salaryMax: s.salaryMax ?? null,
    recruiterLinkedIn: s.recruiterLinkedIn ?? null,
    applyUrl: s.applyUrl ?? null,
    createdAt: doc.createdAt.toISOString(),
    updatedAt: doc.createdAt.toISOString(),
    isSaved: true,
  };
}

export const jobService = {
  toDto,

  /**
   * Internal listings (paginated, always shown) plus a best-effort page of
   * external results (only when searching with a keyword — JSearch has no
   * "browse all" concept — and only when source isn't restricted to
   * "internal"). savedOnly bypasses this entirely and reads bookmarked
   * snapshots instead, since a saved external listing has no other home
   * once it falls off a live search page.
   */
  async search(filters: JobSearchFilters, page: number, limit: number, userId: string) {
    const safePage = Math.max(1, page);
    const safeLimit = Math.min(MAX_PAGE_SIZE, Math.max(1, limit));

    if (filters.savedOnly) {
      const saved = await savedJobRepository.listByUser(userId);
      const start = (safePage - 1) * safeLimit;
      const pageDocs = saved.slice(start, start + safeLimit);
      return {
        items: pageDocs.map(toSavedDto),
        page: safePage,
        limit: safeLimit,
        total: saved.length,
        totalPages: Math.max(1, Math.ceil(saved.length / safeLimit)),
      };
    }

    const postedAfter = datePostedToCutoff(filters.datePosted);
    const includeExternal = filters.source !== "internal" && !!filters.q;

    const [[internalDocs, total], externalRaw, savedIds] = await Promise.all([
      jobRepository.search(
        {
          query: filters.q,
          location: filters.location,
          company: filters.company,
          employmentType: filters.employmentType,
          workMode: filters.workMode,
          experienceLevel: filters.experienceLevel,
          skill: filters.skill,
          source: filters.source,
          salaryMin: filters.salaryMin,
          postedAfter,
        },
        safePage,
        safeLimit,
      ),
      includeExternal ? searchExternalJobs(filters.q!, filters.datePosted) : Promise.resolve([]),
      savedJobRepository.savedJobIds(userId),
    ]);

    // External results can't be filtered server-side by salary (JSearch doesn't expose it as a query
    // param), so the floor is applied here once the salary fields it does return are available.
    const external = externalRaw.filter((j) => meetsSalaryFloor(j, filters.salaryMin));

    const items = [...internalDocs.map(toDto), ...external].map((j) => ({ ...j, isSaved: savedIds.has(j.id) }));

    return {
      items,
      page: safePage,
      limit: safeLimit,
      total,
      totalPages: Math.max(1, Math.ceil(total / safeLimit)),
    };
  },

  async getFilterOptions(): Promise<JobFilterOptions> {
    return jobRepository.distinctFacets();
  },

  /** Toggles a bookmark. Saving needs the job's display fields — for internal jobs we can re-fetch them,
   *  but external (jsearch) jobs only exist as a live API response, so the caller must supply them. */
  async toggleSave(userId: string, jobId: string, jobPayload?: Job): Promise<{ saved: boolean }> {
    const existing = await savedJobRepository.find(userId, jobId);
    if (existing) {
      await savedJobRepository.unsave(userId, jobId);
      return { saved: false };
    }

    let job = jobPayload;
    if (!job && !jobId.startsWith("jsearch:")) {
      const doc = await jobRepository.findById(jobId);
      if (doc) job = toDto(doc);
    }
    if (!job) throw new ValidationError("Job details are required to save this listing.");

    await savedJobRepository.save(userId, jobId, job);
    return { saved: true };
  },

  async getById(id: string): Promise<Job> {
    const doc = await jobRepository.findById(id);
    if (!doc) throw new NotFoundError("Job not found");
    return toDto(doc);
  },

  async listMine(userId: string): Promise<Job[]> {
    const docs = await jobRepository.listByPoster(userId);
    return docs.map(toDto);
  },

  async create(userId: string, input: CreateJobInput): Promise<Job> {
    const doc = await jobRepository.create({
      postedBy: userId,
      title: input.title,
      company: input.company,
      location: input.location ?? null,
      description: input.description,
      skills: input.skills ?? [],
      employmentType: input.employmentType ?? null,
      workMode: input.workMode ?? null,
      experienceLevel: input.experienceLevel ?? null,
      salaryMin: input.salaryMin ?? null,
      salaryMax: input.salaryMax ?? null,
      recruiterLinkedIn: input.recruiterLinkedIn ?? null,
      applyUrl: input.applyUrl ?? null,
    });
    return toDto(doc);
  },

  async update(id: string, userId: string, input: UpdateJobInput): Promise<Job> {
    const doc = await jobRepository.update(id, userId, input);
    if (!doc) throw new NotFoundError("Job not found");
    return toDto(doc);
  },

  async delete(id: string, userId: string): Promise<void> {
    const deleted = await jobRepository.delete(id, userId);
    if (!deleted) throw new NotFoundError("Job not found");
  },
};

File: apps/backend/services/job_service.py
"""Job Board — mirrors apps/api's job.service.ts, including the internal +
external (JSearch) result merge and the saved-jobs bypass path."""

from __future__ import annotations

from dataclasses import dataclass
from datetime import UTC, datetime, timedelta
from typing import Any

from bson import ObjectId

from clients.jsearch_client import search_external_jobs
from core.config import Settings
from core.errors import NotFoundError, ValidationError
from models.job import Job
from models.saved_job import JobSnapshot, SavedJob
from repositories import job_repository, saved_job_repository
from repositories.job_repository import JobSearchFilters as RepoJobSearchFilters
from schemas.job import CreateJobInput, DatePostedFilter, JobDto, JobFilterOptions, LocationOption, UpdateJobInput
from schemas.pagination import PaginatedResult, safe_limit, safe_page, total_pages
from services.location_parsing import matches as location_matches
from services.location_parsing import parse_location

_DATE_POSTED_DELTA = {
    "24h": timedelta(hours=24),
    "7d": timedelta(days=7),
    "30d": timedelta(days=30),
}


@dataclass
class JobSearchFilters:
    q: str | None = None
    country: str | None = None
    state: str | None = None
    city: str | None = None
    company: str | None = None
    employmentType: str | None = None
    workMode: str | None = None
    experienceLevel: str | None = None
    skill: str | None = None
    source: str | None = None
    salaryMin: float | None = None
    datePosted: DatePostedFilter | None = None
    savedOnly: bool = False


def _date_posted_to_cutoff(date_posted: DatePostedFilter | None) -> datetime | None:
    if not date_posted or date_posted == "any":
        return None
    return datetime.now(UTC) - _DATE_POSTED_DELTA[date_posted]


def _meets_salary_floor(job: JobDto, floor: float | None) -> bool:
    if floor is None:
        return True
    ceiling = job.salaryMax if job.salaryMax is not None else job.salaryMin
    return ceiling is not None and ceiling >= floor


def _to_dto(doc: Job) -> JobDto:
    return JobDto(
        id=str(doc.id),
        source="internal",
        postedBy=str(doc.postedBy),
        title=doc.title,
        company=doc.company,
        location=doc.location,
        description=doc.description,
        skills=doc.skills,
        employmentType=doc.employmentType,
        workMode=doc.workMode,
        experienceLevel=doc.experienceLevel,
        salaryMin=doc.salaryMin,
        salaryMax=doc.salaryMax,
        recruiterLinkedIn=doc.recruiterLinkedIn,
        applyUrl=doc.applyUrl,
        createdAt=doc.createdAt.isoformat(),
        updatedAt=doc.updatedAt.isoformat(),
    )


def _to_saved_dto(doc: SavedJob) -> JobDto:
    s = doc.snapshot
    created = doc.createdAt.isoformat()
    return JobDto(
        id=doc.jobId,
        source=s.source,  # type: ignore[arg-type]
        postedBy=None,
        title=s.title,
        company=s.company,
        location=s.location,
        description=s.description,
        skills=s.skills,
        employmentType=s.employmentType,
        workMode=s.workMode,
        experienceLevel=s.experienceLevel,
        salaryMin=s.salaryMin,
        salaryMax=s.salaryMax,
        recruiterLinkedIn=s.recruiterLinkedIn,
        applyUrl=s.applyUrl,
        createdAt=created,
        updatedAt=created,
        isSaved=True,
    )


def _snapshot_from_dto(job: JobDto) -> JobSnapshot:
    return JobSnapshot(
        source=job.source,
        title=job.title,
        company=job.company,
        location=job.location,
        description=job.description,
        skills=job.skills,
        employmentType=job.employmentType,
        workMode=job.workMode,
        experienceLevel=job.experienceLevel,
        salaryMin=job.salaryMin,
        salaryMax=job.salaryMax,
        recruiterLinkedIn=job.recruiterLinkedIn,
        applyUrl=job.applyUrl,
    )


async def search(settings: Settings, filters: JobSearchFilters, page: int, limit: int, user_id: ObjectId) -> PaginatedResult[JobDto]:
    """Internal listings (paginated, always shown) plus a best-effort page of
    external results (only when searching with a keyword — JSearch has no
    "browse all" concept — and only when source isn't restricted to
    "internal"). savedOnly bypasses this entirely and reads bookmarked
    snapshots instead, since a saved external listing has no other home
    once it falls off a live search page."""
    safe_page_ = safe_page(page)
    safe_limit_ = safe_limit(limit)

    if filters.savedOnly:
        saved = await saved_job_repository.list_by_user(user_id)
        start = (safe_page_ - 1) * safe_limit_
        page_docs = saved[start : start + safe_limit_]
        return PaginatedResult[JobDto](
            items=[_to_saved_dto(doc) for doc in page_docs],
            page=safe_page_,
            limit=safe_limit_,
            total=len(saved),
            totalPages=total_pages(len(saved), safe_limit_),
        )

    posted_after = _date_posted_to_cutoff(filters.datePosted)
    include_external = filters.source != "internal" and bool(filters.q)
    has_location_filter = bool(filters.country or filters.state or filters.city)

    location_raws: list[str] | None = None
    if has_location_filter:
        location_raws = await job_repository.raw_locations_matching(filters.country, filters.state, filters.city)

    (internal_docs, total), external_raw, saved_ids = (
        await job_repository.search(
            RepoJobSearchFilters(
                query=filters.q,
                locationRaws=location_raws,
                company=filters.company,
                employmentType=filters.employmentType,  # type: ignore[arg-type]
                workMode=filters.workMode,  # type: ignore[arg-type]
                experienceLevel=filters.experienceLevel,  # type: ignore[arg-type]
                skill=filters.skill,
                source=filters.source,  # type: ignore[arg-type]
                salaryMin=filters.salaryMin,
                postedAfter=posted_after,
            ),
            safe_page_,
            safe_limit_,
        ),
        (await search_external_jobs(settings, filters.q, filters.datePosted) if include_external and filters.q else []),
        await saved_job_repository.saved_job_ids(user_id),
    )

    # External results can't be filtered server-side by salary or structured location (JSearch
    # doesn't expose either as a query param), so both are applied here against the fields it does return.
    external = [
        j
        for j in external_raw
        if _meets_salary_floor(j, filters.salaryMin)
        and (not has_location_filter or location_matches(parse_location(j.location), filters.country, filters.state, filters.city))
    ]

    items = [item.model_copy(update={"isSaved": item.id in saved_ids}) for item in [*[_to_dto(d) for d in internal_docs], *external]]

    return PaginatedResult[JobDto](items=items, page=safe_page_, limit=safe_limit_, total=total, totalPages=total_pages(total, safe_limit_))


async def get_filter_options() -> JobFilterOptions:
    locations, location_triples, companies, skills = await job_repository.distinct_facets()
    location_options = [LocationOption(country=c, state=s, city=ci) for c, s, ci in location_triples]
    return JobFilterOptions(locations=locations, locationOptions=location_options, companies=companies, skills=skills)


async def toggle_save(user_id: ObjectId, job_id: str, job_payload: JobDto | None) -> bool:
    """Toggles a bookmark. Saving needs the job's display fields — for internal jobs we can
    re-fetch them, but external (jsearch) jobs only exist as a live API response, so the
    caller must supply them. Returns True if now saved, False if now unsaved."""
    existing = await saved_job_repository.find(user_id, job_id)
    if existing:
        await saved_job_repository.unsave(user_id, job_id)
        return False

    job = job_payload
    if not job and not job_id.startswith("jsearch:"):
        doc = await job_repository.find_by_id(ObjectId(job_id))
        if doc:
            job = _to_dto(doc)
    if not job:
        raise ValidationError("Job details are required to save this listing.")

    await saved_job_repository.save(user_id, job_id, _snapshot_from_dto(job))
    return True


async def get_by_id(job_id: ObjectId) -> JobDto:
    doc = await job_repository.find_by_id(job_id)
    if not doc:
        raise NotFoundError("Job not found")
    return _to_dto(doc)


async def list_mine(user_id: ObjectId) -> list[JobDto]:
    docs = await job_repository.list_by_poster(user_id)
    return [_to_dto(doc) for doc in docs]


async def create(user_id: ObjectId, input_: CreateJobInput) -> JobDto:
    doc = await job_repository.create(user_id, input_.model_dump())
    return _to_dto(doc)


async def update(job_id: ObjectId, user_id: ObjectId, input_: UpdateJobInput) -> JobDto:
    updates: dict[str, Any] = input_.model_dump(exclude_unset=True)
    doc = await job_repository.update(job_id, user_id, updates)
    if not doc:
        raise NotFoundError("Job not found")
    return _to_dto(doc)


async def delete(job_id: ObjectId, user_id: ObjectId) -> None:
    deleted = await job_repository.delete(job_id, user_id)
    if not deleted:
        raise NotFoundError("Job not found")
